File: src/test/resources/PalindromeList.java
import com.jalgoarena.type.ListNode;

public class Solution {
    /**
     * @param head Linked List to check if it's palindrome
     * @return  Indicates if input linked list is palindrome
     */
    public boolean isPalindrome(ListNode head) {
        ListNode reversed = reverseAndClone(head);
        return isEqual(head, reversed);
    }

    private ListNode reverseAndClone(ListNode node) {
        ListNode head = null;
        while (node != null) {
            ListNode n = new ListNode(node.value);
            n.next = head;
            head = n;
            node = node.next;
        }
        return head;
    }

    private boolean isEqual(ListNode one, ListNode two) {
        while (one != null && two != null) {
            if (one.value != two.value) {
                return false;
            }
            one = one.next;
            two = two.next;
        }
        return one == null && two == null;
    }
}

File: src/test/resources/PalindromeList.kt
import com.jalgoarena.type.ListNode
import com.jalgoarena.type.*

class PalindromeList {
    /**
     * @param head Linked List to check if it's palindrome
     * *
     * @return  Indicates if input linked list is palindrome
     */
    fun isPalindrome(head: ListNode?): Boolean {
        val reversed = reverseAndClone(head)
        return isEqual(head, reversed)
    }

    private fun reverseAndClone(listNode: ListNode?): ListNode? {
        var node = listNode
        var head: ListNode? = null
        while (node != null) {
            val n = ListNode(node.value)
            n.next = head
            head = n
            node = node.next
        }
        return head
    }

    private fun isEqual(first: ListNode?, second: ListNode?): Boolean {
        var one = first
        var two = second
        while (one != null && two != null) {
            if (one.value != two.value) {
                return false
            }
            one = one.next
            two = two.next
        }
        return one == null && two == null
    }
}
